certificates.go

```diff
@@ -0,0 +1,168 @@
+	err := c.withSession(func(session *pkcs11Session) (err error) {
+		if id == nil && label == nil && serial == nil {
+			return errors.New("id, label and serial cannot all be nil")
+		}
+
+		var template []*pkcs11.Attribute
+
+		if id != nil {
+			template = append(template, pkcs11.NewAttribute(pkcs11.CKA_ID, id))
+		}
+		if label != nil {
+			template = append(template, pkcs11.NewAttribute(pkcs11.CKA_LABEL, label))
+		}
+		if serial != nil {
+			derSerial, err := asn1.Marshal(serial)
+			if err != nil {
+				return errors.WithMessage(err, "failed to encode serial")
+			}
+
+			template = append(template, pkcs11.NewAttribute(pkcs11.CKA_SERIAL_NUMBER, derSerial))
+		}
+
+		template = append(template, pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_CERTIFICATE))
+
+		if err = session.ctx.FindObjectsInit(session.handle, template); err != nil {
+			return err
+		}
+		defer func() {
+			finalErr := session.ctx.FindObjectsFinal(session.handle)
+			if err == nil {
+				err = finalErr
+			}
+		}()
+
+		handles, _, err := session.ctx.FindObjects(session.handle, 1)
+		if err != nil {
+			return err
+		}
+		if len(handles) == 0 {
+			return nil
+		}
+
+		attributes := []*pkcs11.Attribute{
+			pkcs11.NewAttribute(pkcs11.CKA_VALUE, 0),
+		}
+
+		if attributes, err = session.ctx.GetAttributeValue(session.handle, handles[0], attributes); err != nil {
+			return err
+		}
+
+		cert, err = x509.ParseCertificate(attributes[0].Value)
+		return err
+	})
+
+	return cert, err
+}
+
+// ImportCertificate imports a certificate onto the token. The id parameter is used to
+// set CKA_ID and must be non-nil.
+func (c *Context) ImportCertificate(id []byte, certificate *x509.Certificate) error {
+	if c.closed.Get() {
+		return errClosed
+	}
+
+	if err := notNilBytes(id, "id"); err != nil {
+		return err
+	}
+
+	template, err := NewAttributeSetWithID(id)
+	if err != nil {
+		return err
+	}
+	return c.ImportCertificateWithAttributes(template, certificate)
+}
+
+// ImportCertificateWithLabel imports a certificate onto the token.  The id and label parameters are used to
+// set CKA_ID and CKA_LABEL respectively and must be non-nil.
+func (c *Context) ImportCertificateWithLabel(id []byte, label []byte, certificate *x509.Certificate) error {
+	if c.closed.Get() {
+		return errClosed
+	}
+
+	if err := notNilBytes(id, "id"); err != nil {
+		return err
+	}
+	if err := notNilBytes(label, "label"); err != nil {
+		return err
+	}
+
+	template, err := NewAttributeSetWithIDAndLabel(id, label)
+	if err != nil {
+		return err
+	}
+	return c.ImportCertificateWithAttributes(template, certificate)
+}
+
+// ImportCertificateWithAttributes imports a certificate onto the token. After this function returns, template
+// will contain the attributes applied to the certificate. If required attributes are missing, they will be set to a
+// default value.
+func (c *Context) ImportCertificateWithAttributes(template AttributeSet, certificate *x509.Certificate) error {
+	if certificate == nil {
```

Comment:
Is there a missing `c.closed.Get()` check here?

Reply:
Good spot. I've added a test to catch this and the fix. I've also opened #55, because I've noticed some other public functions are not tested.